grpc/cpp-cubic/CobaltGRPC.cmake

```diff
@@ -5,139 +5,128 @@
+            "${grpc_SOURCE_DIR}/third_party/protobuf/src"
+            "${grpc_SOURCE_DIR}/third_party/googleapis")
     endif()
 
-    # If we are cross-compiling, add the native grpc dir to the CMAKE_PREFIX_PATH
-    # so that the grpc build can find it using find_program
+    get_filename_component(CUBIC_PROTO_DIR "${cubic_client_SOURCE_DIR}/.." ABSOLUTE)
+    list(APPEND PROTOC_INCLUDE_LIST "${CUBIC_PROTO_DIR}")
+
+    foreach(protofile ${RUN_PROTOC_PROTOS})
+        get_filename_component(absolute_proto "${protofile}" ABSOLUTE)
+        get_filename_component(base_name "${absolute_proto}" NAME_WE)
+        get_filename_component(proto_path "${absolute_proto}" PATH)
+
+        list(APPEND PROTOC_INCLUDE_LIST "${proto_path}")
+        list(APPEND ALL_PROTOS "${absolute_proto}")
+
+        list(APPEND ${RUN_PROTOC_RESULT_FILES}
+                "${OUTDIR}/${base_name}.grpc.pb.cc"
+                "${OUTDIR}/${base_name}.grpc.pb.h"
+                "${OUTDIR}/${base_name}.pb.cc"
+                "${OUTDIR}/${base_name}.pb.h")
+    endforeach()
+
+    # Uniquify the include path list and format it properly
+    list(REMOVE_DUPLICATES PROTOC_INCLUDE_LIST)
+    if(WIN32)
+        # For windows, instead of providing a single delimited list,
+        # we will just add the include flag for each path.
+        string(REPLACE ";" " -I " PROTOC_INCLUDES "${PROTOC_INCLUDE_LIST}")
+        separate_arguments(PROTOC_INCLUDES WINDOWS_COMMAND "${PROTOC_INCLUDES}")
+    else()
+        string(REPLACE ";" ":" PROTOC_INCLUDES "${PROTOC_INCLUDE_LIST}")
+    endif()
+
+    # To enable cross-compiling, we allow users to specify a native
+    # grpc installation, which we can use to run protoc. The version
+    # of this native grpc should be the same as the one used for
+    # cross compiling.
+    set(_PROTOC_CMD $<TARGET_FILE:protoc>)
+    set(_GRPC_CPP_PLUGIN $<TARGET_FILE:grpc_cpp_plugin>)
     if(EXISTS ${GRPC_NATIVE_DIR})
-        set(CMAKE_PREFIX_PATH ${CMAKE_PREFIX_PATH} ${GRPC_NATIVE_DIR})
+        get_filename_component(absolute_grpc_native "${GRPC_NATIVE_DIR}" ABSOLUTE)
+        set(_PROTOC_CMD "${absolute_grpc_native}/bin/protoc")
+        set(_GRPC_CPP_PLUGIN "${absolute_grpc_native}/bin/grpc_cpp_plugin")
     endif()
 
-    # Make gRPC available to the project.
-    FetchContent_MakeAvailable(gRPC)
-endif()
+    add_custom_command(
+        OUTPUT ${${RUN_PROTOC_RESULT_FILES}}
+        COMMAND ${_PROTOC_CMD}
+        ARGS -I ${PROTOC_INCLUDES}
+            --grpc_out=${RUN_PROTOC_OUTPUT_DIR}
+            --cpp_out=${RUN_PROTOC_OUTPUT_DIR}
+            --plugin=protoc-gen-grpc=${_GRPC_CPP_PLUGIN}
+            ${ALL_PROTOS}
+        DEPENDS ${ALL_PROTOS})
+
+    # Make sure the output variable specified makes it back to the
+    # parent scope.
+    set(${RUN_PROTOC_RESULT_FILES} ${${RUN_PROTOC_RESULT_FILES}} PARENT_SCOPE)
+
+endfunction(run_protoc)
 
-# Check if we have already defined the run_protoc function
-if(NOT COMMAND run_protoc)
```

Comment:
Just for my education, why do we no longer want to check whether `run_protoc` is already defined?

Reply:
I think I decided at some point that I didn't care if it was defined or not, I wanted to use my `run_protoc` command. (I may have also decided it was very unlikely for the function to already be defined).